{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e57ab144",
   "metadata": {},
   "source": [
    "### Introduction\n",
    "Coming soon ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "id": "952d7070",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import modules and packages\n",
    "import copy\n",
    "from pomegranate import *\n",
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import (confusion_matrix, precision_recall_curve, auc, roc_auc_score,\n",
    "                             roc_curve, recall_score, classification_report, f1_score,\n",
    "                             precision_recall_fscore_support, precision_score, accuracy_score, matthews_corrcoef)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "id": "11130f56",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_model_from_json(filename=None, verbose=False):\n",
    "    \"\"\"\n",
    "    Reads and returns model data from JSON file.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    filename : str\n",
    "               Path to a saved model in JSON format\n",
    "    verbose  : bool, optional, default False\n",
    "               When True, prints progress to STDOUT\n",
    "    \n",
    "    Returns\n",
    "    ----------\n",
    "    dict\n",
    "    JSON data as dictionary\n",
    "    \"\"\"\n",
    "    \n",
    "    with open(filename, 'r') as f:\n",
    "        modeldata = json.load(f)\n",
    "        #print(modeldata['edges'][2])\n",
    "        if verbose: print(\"Model read from:\", filename)\n",
    "        return(modeldata)\n",
    "\n",
    "\n",
    "\n",
    "def train_hmm(distribution=None, n_components=None, state_names=None, X=None, labels=None, algorithm=None, n_jobs=None):\n",
    "    \"\"\"\n",
    "    Helper function to train and bake Pomogranate HMM model. Returns baked Pomogranate HMM model.\n",
    "    Make sure you import the pomogranate package as-\n",
    "    from pomegranate import *\n",
    "    \n",
    "    This function uses (docstrings are appended below)-\n",
    "    pomogranate.HiddenMarkovModel.from_samples\n",
    "    pomogranate.HiddenMarkovModel.bake\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    \"\"\"\n",
    "    # Example:\n",
    "    model = HiddenMarkovModel.from_samples(distribution=DiscreteDistribution, n_components=2,\n",
    "                                         state_names=['F','T'],\n",
    "                                         X=train_seqs, labels=train_labels,\n",
    "                                         algorithm='labeled',\n",
    "                                         n_jobs=2)\n",
    "    \"\"\"\n",
    "    \n",
    "    model = HiddenMarkovModel.from_samples(distribution=distribution, n_components=n_components,\n",
    "                                         state_names=state_names,\n",
    "                                         X=X, labels=labels,\n",
    "                                         algorithm=algorithm,\n",
    "                                         n_jobs=n_jobs)\n",
    "    model.bake()\n",
    "    return(model)\n",
    "# Append docstring\n",
    "train_hmm.__doc__ += \"\\nHiddenMarkovModel.from_samples\\n\\n\" + HiddenMarkovModel.from_samples.__doc__\n",
    "train_hmm.__doc__ += \"\\n\\nHiddenMarkovModel.bake\\n\\n\" + HiddenMarkovModel.bake.__doc__\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "id": "8c8375f8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def validate(model=None, states=None, true_seqs=None, true_labels=None):\n",
    "    \"\"\"\n",
    "    Scan \n",
    "    \"\"\"\n",
    "    y_true_all = []\n",
    "    y_pred_all = []\n",
    "    F_predictedAs_T_fraction = []\n",
    "    T_predictedAs_F_fraction = []\n",
    "    recall = []\n",
    "    precision = []\n",
    "    accuracy = []\n",
    "    f1 = []\n",
    "    for i in true_seqs.index:\n",
    "        y_true = [int(ch=='T') for ch in true_labels[i]]\n",
    "        y_true_all += copy.copy(y_true)\n",
    "        y_pred = model.predict(list(true_seqs[i]))\n",
    "        y_pred_all += copy.copy(y_pred)\n",
    "        #print([int(ch=='T') for ch in true_labels[i]])\n",
    "        #print(\"\".join([states[i] for i in pred_labels]))\n",
    "        #print(pred_labels)\n",
    "        #print(i,\"\\n\",true_seqs[i],\"\\n\",true_labels[i],\"\\n\",\"\".join([states[i] for i in y_pred]))\n",
    "\n",
    "        \"\"\"\n",
    "        if 1:#i == \">tr|A0A0R2L6D1|tig\":\n",
    "            confmat = confusion_matrix(y_true,y_pred)\n",
    "            print(confmat)\n",
    "            print(recall_score(y_true,y_pred))\n",
    "            print(precision_score(y_true,y_pred))\n",
    "            print(accuracy_score(y_true,y_pred))\n",
    "            #print(i, true_seqs[i], true_labels[i])\n",
    "            print(\"---\")\n",
    "        \"\"\"\n",
    "\n",
    "        confmat = confusion_matrix(y_true,y_pred)\n",
    "        #print(confmat)\n",
    "        #print(confmat.shape)\n",
    "        if confmat.shape[0] == 1 and confmat.shape[1] == 1:\n",
    "            F_predictedAs_T = 0\n",
    "            T_predictedAs_F = 0\n",
    "        else:\n",
    "            F_predictedAs_T = confmat[0,1]\n",
    "            T_predictedAs_F = confmat[1,0]\n",
    "            num_res = np.sum(confmat)\n",
    "            F_predictedAs_T_fraction.append(F_predictedAs_T/num_res)\n",
    "            T_predictedAs_F_fraction.append(T_predictedAs_F/num_res)\n",
    "        recall.append(recall_score(y_true,y_pred))\n",
    "        precision.append(precision_score(y_true,y_pred))\n",
    "        accuracy.append(accuracy_score(y_true,y_pred))\n",
    "        f1.append(f1_score(y_true,y_pred))\n",
    "\n",
    "        #if accuracy_score(y_true,y_pred) <= 0.75:\n",
    "        #    print(i,\"\\n\",true_seqs[i],\"\\n\",true_labels[i],\"\\n\",\"\".join([states[i] for i in y_pred]))\n",
    "    \n",
    "\n",
    "    #print(np.mean(recall), np.mean(precision), np.mean(accuracy), np.mean(f1), np.mean(F_predictedAs_T_fraction), np.mean(T_predictedAs_F_fraction))\n",
    "    #print(np.std(recall),    np.std(precision),    np.std(accuracy), np.std(f1), np.std(F_predictedAs_T_fraction), np.std(T_predictedAs_F_fraction))\n",
    "\n",
    "    \"\"\"\n",
    "    print(\"Histograms\\n\")\n",
    "    print(\"Recall:\")\n",
    "    H,xe = histogram1d(data=recall,binsize=0.1,xmin=0.0,xmax=1.0)\n",
    "    print(H,\"\\n\",xe,\"\\n\")\n",
    "    print(\"Precision:\")\n",
    "    H,xe = histogram1d(data=precision,binsize=0.1,xmin=0.0,xmax=1.0)\n",
    "    print(H,\"\\n\",xe,\"\\n\")\n",
    "    print(\"Accuracy:\")\n",
    "    H,xe = histogram1d(data=accuracy,binsize=0.1,xmin=0.0,xmax=1.0)\n",
    "    print(H,\"\\n\",xe,\"\\n\")\n",
    "    print(\"F1 score:\")\n",
    "    H,xe = histogram1d(data=f1,binsize=0.1,xmin=0.0,xmax=1.0)\n",
    "    print(H,\"\\n\",xe,\"\\n\")\n",
    "    print(\"F_predictedAs_T_fraction:\")\n",
    "    H,xe = histogram1d(data=F_predictedAs_T_fraction,binsize=0.1,xmin=0.0,xmax=1.0)\n",
    "    print(H,\"\\n\",xe,\"\\n\")\n",
    "    print(\"T_predictedAs_F_fraction:\")\n",
    "    H,xe = histogram1d(data=T_predictedAs_F_fraction,binsize=0.1,xmin=0.0,xmax=1.0)\n",
    "    print(H,\"\\n\",xe,\"\\n\")\n",
    "\n",
    "    print(\"Total num seq for validation:\", len(true_seqs), len(recall))\n",
    "    print(\"Total num seq for F_predictedAs_T:\", len(F_predictedAs_T_fraction))\n",
    "    print(\"Total num seq for T_predictedAs_F:\", len(T_predictedAs_F_fraction))\n",
    "\n",
    "    print()\n",
    "    \"\"\"\n",
    "    \n",
    "    \"\"\"\n",
    "    print(len(y_true_all),len(y_pred_all))\n",
    "    print()\n",
    "    print(\"Confusion Matrix:\")\n",
    "    print(confusion_matrix(y_true_all,y_pred_all))\n",
    "    print(\"Recall:\",recall_score(y_true_all,y_pred_all))\n",
    "    print(\"Precision:\",precision_score(y_true_all,y_pred_all))\n",
    "    print(\"Accuracy:\",accuracy_score(y_true_all,y_pred_all))\n",
    "    print(\"MCC:\",matthews_corrcoef(y_true_all,y_pred_all))\n",
    "    print(\"AUC:\",roc_auc_score(y_true_all,y_pred_all))\n",
    "    print(\"F1 score:\",f1_score(y_true_all,y_pred_all))\n",
    "    \"\"\"\n",
    "\n",
    "    return {\"ConfusionMatrix\":confusion_matrix(y_true_all,y_pred_all),\n",
    "                    \"Recall\":recall_score(y_true_all,y_pred_all),\n",
    "                    \"Precision\":precision_score(y_true_all,y_pred_all),\n",
    "                    \"Accuracy\":accuracy_score(y_true_all,y_pred_all),\n",
    "                    \"MCC\":matthews_corrcoef(y_true_all,y_pred_all),\n",
    "                    \"AUC\":roc_auc_score(y_true_all,y_pred_all),\n",
    "                    \"F1score\":f1_score(y_true_all,y_pred_all)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "id": "1112a543",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Temporary functions\n",
    "def histogram1d(data=None, binsize=None, xmin=None, xmax=None):\n",
    "    if xmin is None and xmax is None:\n",
    "        xmin = min(data)\n",
    "        xmax = max(data)\n",
    "    xedges = np.arange(xmin,xmax+binsize,binsize)\n",
    "    H, xe = np.histogram(data,bins=xedges)\n",
    "    #print(\"Value range:\",xmin,xmax)\n",
    "    return(H,xe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "id": "9b76f4e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def translate(seq=None,translation_dict=None):\n",
    "    \"\"\"\n",
    "    Translates sequence of characters based on the provided translation dictionary.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    seq              : str, default None\n",
    "                       Sequence of characters to be translated\n",
    "    translation_dict : dict, default None\n",
    "                       Dictionary to use for translation of 'seq'.\n",
    "                       All characters in the input sequence must be present in the keys of the translation dictionary.\n",
    "    \n",
    "    Returns\n",
    "    ----------\n",
    "    str\n",
    "    Translated sequence\n",
    "    \n",
    "    \"\"\"\n",
    "    if translation_dict:\n",
    "        return(\"\".join([translation_dict[ch] for ch in seq]))\n",
    "    else:\n",
    "        return(\"\".join([ch for ch in seq]))\n",
    "\n",
    "    \n",
    "def get_seq_dictionary(datadir=None, filelist=None, header_regex_str=None, translation_dict=None, verbose=False):\n",
    "    \"\"\"\n",
    "    Translates sequence of characters based on the provided translation dictionary.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    datadir          : str, default None\n",
    "                       Path to directory that contains files containing sequences in FASTA format (shown below).\n",
    "                       > sequence name 1\n",
    "                       ALKDSFJOSDIFUSDCOSDNWEIFUCSNDJ\n",
    "                       > sequence name 2\n",
    "                       OSIFJODSICVJIWDCNSIDCIWF\n",
    "                       > sequence name 3\n",
    "                       LAKSDJFOWIEROSICJOISFJLSKDFD\n",
    "    filelist         : list, default None\n",
    "                       List of file names (str) to read from 'datadir'\n",
    "    header_regex_str : str, default None\n",
    "                       The header is parsed according to this regex string and first match group is used at the key.\n",
    "                       If None then the full header line (except the last new line) is used as the key.\n",
    "    translation_dict : dict, default None\n",
    "                       Dictionary to use for translation of 'seq'.\n",
    "                       All characters in the input sequence must be present in the keys of the translation dictionary.\n",
    "    verbose          : bool, optional, default False\n",
    "                       When True, prints information to STDOUT\n",
    "\n",
    "    \n",
    "    Returns\n",
    "    ----------\n",
    "    dict\n",
    "    Dictionary of sequences.\n",
    "    Keys are sequence headers read from FASTA formatted sequences.\n",
    "    Values are sequences translated using 'traslation_dict'.\n",
    "    If a duplicate sequence header is found then it is replaced by the one found later.\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    if header_regex_str is None:\n",
    "        header_regex_str = '(.*)\\n'\n",
    "\n",
    "    seq = {}\n",
    "    for filename in filelist:\n",
    "        try:\n",
    "            f = open(datadir+\"/\"+filename,'r')\n",
    "        except:\n",
    "            if verbose:\n",
    "                print(\"Error while opening file\", datadir+\"/\"+filename)\n",
    "            continue\n",
    "\n",
    "        for line in f.readlines():\n",
    "            if line.startswith('>'):\n",
    "                matchObj = re.match(header_regex_str,line,re.S)\n",
    "                header = matchObj.group(1)\n",
    "                seq[header] = \"\"\n",
    "            else:\n",
    "                seq[header] += translate(line.replace('\\n',''),translation_dict=translation_dict)\n",
    "        f.close()\n",
    "    return(seq)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03521191",
   "metadata": {},
   "source": [
    "<h3>Define input variables</h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "id": "9e8b6133",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Data directory\n",
    "datadir = \"pakTailSeqs\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "id": "98ba618b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### List of tail and full length sequence files\n",
    "\n",
    "tail_seq_files = [\n",
    "#'adk_cdhit90_tails.fasta',\n",
    "#'hbha_cdhit90_tails.fasta',\n",
    "#'hc1_cdhit90_tails.fasta',\n",
    "#'hc2_cdhit90_tails.fasta',\n",
    "'hu_cdhit90_tails.fasta'\n",
    "#'ku_cdhit90_tails.fasta'\n",
    "#'l22_cdhit90_tails.fasta',\n",
    "#'rbr_cdhit90_tails.fasta'#? Manually identified tails are probably not correct PAK tails.\n",
    "#'sig_cdhit90_tails.fasta'#? Manually identified tails are probably not correct PAK tails.\n",
    "#'tig_cdhit90_tails.fasta',\n",
    "#'topo1_cdhit90_tails.fasta',\n",
    "#'topo3_cdhit90_tails.fasta'\n",
    "]\n",
    "\n",
    "full_seq_files = [\n",
    "#'adk_cdhit90.fasta',\n",
    "#'hbha_cdhit90.fasta',\n",
    "#'hc1_cdhit90.fasta',\n",
    "#'hc2_cdhit90.fasta',\n",
    "'hu_cdhit90.fasta'\n",
    "#'ku_cdhit90.fasta'\n",
    "#'l22_cdhit90.fasta',\n",
    "#'rbr_cdhit90.fasta'#? Manually identified tails are probably not correct PAK tails.\n",
    "#'sig_cdhit90.fasta'#? Manually identified tails are probably not correct PAK tails.\n",
    "#'tig_cdhit90.fasta',\n",
    "#'topo1_cdhit90.fasta',\n",
    "#'topo3_cdhit90.fasta'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "id": "2678b533",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define observed states for HMM\n",
    "#\n",
    "# Amino acid residue codes (i.e. observed states)\n",
    "# Amino acids P, A, K and T are most common in manually identified PAK-tails in HU proteins.\n",
    "# Below are the frequencies.\n",
    "# A 0.329\n",
    "# K 0.264\n",
    "# T 0.184\n",
    "# P 0.067\n",
    "# Proline is distinctly infrequent compared to A, K and T,\n",
    "# but PAK-tail analysis described by Khare et al* indicates its importance in the dynamics and conformation of the tails.\n",
    "# Hence Proline is considered as a separate category.\n",
    "#\n",
    "# Categories: This categorization will help in discriminating P, A, K, T and other amino acids (O).\n",
    "# Taking P, A, K and T in one group leads to poly-A or poly-P stretches to be labeled as PAK-tails.\n",
    "# Keeping P, A, K and T as separate categories might help in learning their combinations.\n",
    "# Categories are assigned as follows.\n",
    "# P\n",
    "# A\n",
    "# K\n",
    "# T\n",
    "# O: OTHERS (D,E,G,A,V,L,I,M,C,N,Q,F,Y,W,P,S,T)\n",
    "#\n",
    "# *Conformational heterogeneity in tails of DNA-binding proteins is augmented by proline containing repeats; Khare et al 2017\n",
    "\n",
    "# Amino acid code single letter\n",
    "aa_code1 = ['A','C','D','E','F','G','H','I','K','L','M','N','P','Q','R','S','T','V','W','X','Y']\n",
    "\n",
    "# Amino acid categories i.e. the observed states for HMM\n",
    "aa_code_cat = ['P','A','K','T','O']\n",
    "\n",
    "# Amino acid category mapping dictionary\n",
    "aa_code_cat_dict = {'P':'P',\n",
    "                     'A':'A',\n",
    "                     'K':'K',\n",
    "                     'T':'T',\n",
    "                     'C':'O', 'D':'O', 'E':'O', 'F':'O', 'G':'O', 'H':'O', 'I':'O', 'L':'O',\n",
    "                     'M':'O', 'N':'O', 'Q':'O', 'R':'O', 'S':'O', 'V':'O', 'W':'O', 'X':'O', 'Y':'O'}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "id": "d3e66746",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Define hidden states for HMM. Two states here, Folded and Tail.\n",
    "states = ['F','T']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "id": "315925ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Number of hidden states for HMM\n",
    "n_states = len(states)\n",
    "\n",
    "# Numger of observed staets for HMM\n",
    "n_states_observed = len(aa_code_cat)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb070e2a",
   "metadata": {},
   "source": [
    "<h3>Prepare input data</h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "id": "a59d2f36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of tail sequences: 59 \n",
      "Number of all (full) sequences: 1353\n"
     ]
    }
   ],
   "source": [
    "seq_tail = get_seq_dictionary(datadir=datadir, filelist=tail_seq_files, header_regex_str='>(.+).*\\n', translation_dict=aa_code_cat_dict, verbose=True)\n",
    "seq_full = get_seq_dictionary(datadir=datadir, filelist=full_seq_files, header_regex_str='>(.+?)\\s.*\\n', translation_dict=aa_code_cat_dict, verbose=True)\n",
    "\n",
    "print(\"Number of tail sequences:\",len(seq_tail),\"\\nNumber of all (full) sequences:\",len(seq_full))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "id": "47c6881b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tail seqs: (59,) \n",
      "Full seqs: (1353,)\n",
      "tr|A0A0T9L3R4|A0A0T9L3R4_MYCTX    AOOAPAOOOOPAAOPOAAOTOAOOPOAPAAKKAKOAKOOKOOKOOK...\n",
      "tr|A0A077M8Y5|A0A077M8Y5_9MICO    AOPOKOPKTAOTOOOAPAOOATATATKAAAKKTTTAKOTTAAKKTT...\n",
      "tr|W7SIK7|W7SIK7_9PSEU            OOTOKOPOOTAAAAKPAAOPAAKATAKOOATKPAOOATAAKATATO...\n",
      "tr|H0K1H8|H0K1H8_9PSEU            OOAKKOPKATTTKOOTTOTOOTTAOTAKTAATOPTOTOOTTOOTOA...\n",
      "tr|A0A0H5CLV3|A0A0H5CLV3_9PSEU    OOTKKOPKOAATOPAAAAPAAOATATOTTATOATTATOTTOAAAOT...\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "### Convert data into Pandas objects\n",
    "tail = pd.Series(seq_tail)\n",
    "full = pd.Series(seq_full)\n",
    "print(\"Tail seqs:\",tail.shape,\"\\nFull seqs:\",full.shape)\n",
    "print(tail.head())\n",
    "#print(full.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "id": "a5f3ecff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Columns: Index(['FULL', 'TAIL', 'FULLLEN', 'TAILLEN', 'FOLDLEN', 'FULLLABELS',\n",
      "       'HEADER'],\n",
      "      dtype='object')\n",
      "Shape (all seqs): (1353, 7)\n",
      "Shape (tail seqs): (59, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FULL</th>\n",
       "      <th>TAIL</th>\n",
       "      <th>FULLLEN</th>\n",
       "      <th>TAILLEN</th>\n",
       "      <th>FOLDLEN</th>\n",
       "      <th>FULLLABELS</th>\n",
       "      <th>HEADER</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OOKAOOOOOOTOKOOOOOOOATAAOOOOOOTOOOAOOKOOOOTOTO...</td>\n",
       "      <td>AOAOOOPOOOPAOKOOOATOAAKKAAOKKAPOKKAOAKKAATKAPA...</td>\n",
       "      <td>200</td>\n",
       "      <td>111</td>\n",
       "      <td>89</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "      <td>sp|O33125|DBH_MYCLE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OOKOOOOOAOAOKAOOTKOOAAOAOOAOOOOOOAOOKOKOOOOOAO...</td>\n",
       "      <td></td>\n",
       "      <td>91</td>\n",
       "      <td>0</td>\n",
       "      <td>91</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "      <td>sp|P02348|DBH5_RHILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OOKOOOOOAOAAOAOOPKAOAOOAOOAOOOOOTOAOKAOOOOOOOO...</td>\n",
       "      <td></td>\n",
       "      <td>90</td>\n",
       "      <td>0</td>\n",
       "      <td>90</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "      <td>sp|P05384|DBHB_PSEAE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OOKOOOOOKOAAOAOOOKAAAOOAOOAOOAOOTOOOKOOOOOAOOO...</td>\n",
       "      <td></td>\n",
       "      <td>90</td>\n",
       "      <td>0</td>\n",
       "      <td>90</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "      <td>sp|P0A1R8|DBHB_SALTY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OTKKOOOOOOAKKAOAKKKOOKOOOOTOOOTOTOAOAKOOKOOOOO...</td>\n",
       "      <td></td>\n",
       "      <td>90</td>\n",
       "      <td>0</td>\n",
       "      <td>90</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "      <td>sp|P36206|DBH_THEMA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                FULL  \\\n",
       "0  OOKAOOOOOOTOKOOOOOOOATAAOOOOOOTOOOAOOKOOOOTOTO...   \n",
       "1  OOKOOOOOAOAOKAOOTKOOAAOAOOAOOOOOOAOOKOKOOOOOAO...   \n",
       "2  OOKOOOOOAOAAOAOOPKAOAOOAOOAOOOOOTOAOKAOOOOOOOO...   \n",
       "3  OOKOOOOOKOAAOAOOOKAAAOOAOOAOOAOOTOOOKOOOOOAOOO...   \n",
       "4  OTKKOOOOOOAKKAOAKKKOOKOOOOTOOOTOTOAOAKOOKOOOOO...   \n",
       "\n",
       "                                                TAIL  FULLLEN  TAILLEN  \\\n",
       "0  AOAOOOPOOOPAOKOOOATOAAKKAAOKKAPOKKAOAKKAATKAPA...      200      111   \n",
       "1                                                          91        0   \n",
       "2                                                          90        0   \n",
       "3                                                          90        0   \n",
       "4                                                          90        0   \n",
       "\n",
       "   FOLDLEN                                         FULLLABELS  \\\n",
       "0       89  FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...   \n",
       "1       91  FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...   \n",
       "2       90  FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...   \n",
       "3       90  FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...   \n",
       "4       90  FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...   \n",
       "\n",
       "                 HEADER  \n",
       "0   sp|O33125|DBH_MYCLE  \n",
       "1  sp|P02348|DBH5_RHILE  \n",
       "2  sp|P05384|DBHB_PSEAE  \n",
       "3  sp|P0A1R8|DBHB_SALTY  \n",
       "4   sp|P36206|DBH_THEMA  "
      ]
     },
     "execution_count": 269,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Make dataframe containing tails and full length sequences\n",
    "### Create columns containing label sequences\n",
    "df = pd.concat([full, tail], axis=1, sort=True)\n",
    "\n",
    "df.columns=['FULL','TAIL']\n",
    "df = df.replace(np.nan, '', regex=True)\n",
    "df['FULLLEN'] = df['FULL'].str.len()\n",
    "df['TAILLEN'] = df['TAIL'].str.len()\n",
    "df['FOLDLEN'] = df['FULLLEN'] - df['TAILLEN']\n",
    "#df['FoldLabels'] = df['FoldLen'].apply(lambda x: 'F'*x)\n",
    "#df['TailLabels'] = df['TailLen'].apply(lambda x: 'T'*x)\n",
    "df['FULLLABELS'] = df['FOLDLEN'].apply(lambda x: 'F'*x) + df['TAILLEN'].apply(lambda x: 'T'*x)\n",
    "#df['Tally'] = df['FullLabels'].str.len()\n",
    "#df.to_csv(path_or_buf=\"test.csv\",sep=\",\")\n",
    "\n",
    "df['HEADER'] = df.index\n",
    "df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "print(\"Columns:\",df.columns)\n",
    "print(\"Shape (all seqs):\",df.shape)\n",
    "#print(df[df['Tail'] != ''].head())\n",
    "print(\"Shape (tail seqs):\",df[df['TAIL'] != ''].shape)\n",
    "df.head()\n",
    "#df.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "id": "2dfd6e7e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "668     OOKAOOOOAOAOKAOOTOOOAKKAOOAOOAOOOOAOKAOOTOOOOO...\n",
      "334     OOKKOOAOOOAOOOOTOKAOOOOOOOOOOOKOOOTOTKOOOOOOAO...\n",
      "17      OOKOOOOOOOATTAOOOKTOATAAOOAOOTOOOOTOAOOOOOAOTO...\n",
      "1062    OOKAOOOOAOAOKOOOOOOAAOAOOOOOOAOOOAOOAOOOOOOTOO...\n",
      "740     OOKOOOOOAOAAOAOOOKAOAKKAOOAOOTOOTOAOKAOOKOOOOO...\n",
      "Name: FULL, dtype: object\n",
      "390    OTTTPPOOOPTOOOOAAOKOOOPOKTOAOOOOOOOKOOOOAKKPOA...\n",
      "789    OOTOOKOOOKKOOKTOOOAOOTOOOOOOKOAAOKTOTKAOAOKAOO...\n",
      "126    OOKTOOOOOOOTKAOOOKAOAOOAOOAOOOAOTTTOKOKOTOTOOO...\n",
      "542    OOKTOOOOAOAOKOOOOKKOATKAOOOOOOTOOOOOKOOOKOOOOO...\n",
      "553    OOKTOOOOKOOOOTOOTKKOAOAAOOAOOOAOTOAOKOOOKOOOOO...\n",
      "Name: FULL, dtype: object\n",
      "(811,) (542,)\n",
      "(811,) (542,)\n",
      "# of seqs with tails in train_seqs 36\n",
      "# of seqs with tails in test_seqs 23\n",
      "\n",
      "The following ratios should be as similar as possible (Adjust random seed accordingly):\n",
      "Train seqs/tails in train seqs = 22.52777777777778\n",
      "Test seqs/tails in test seqs = 23.565217391304348\n"
     ]
    }
   ],
   "source": [
    "### Split input data into train and test sets\n",
    "\n",
    "### Standard way of train test splitting:\n",
    "### X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)\n",
    "### We consider amino acid sequences as X and label sqeuences as y\n",
    "\n",
    "train_seqs, test_seqs, train_labels, test_labels = train_test_split(df.FULL, df.FULLLABELS, test_size=0.4, random_state=61)\n",
    "\n",
    "print(train_seqs.head())\n",
    "print(test_seqs.head())\n",
    "print(train_seqs.shape, test_seqs.shape)\n",
    "print(train_labels.shape, test_labels.shape)\n",
    "#print(train_seqs.head())\n",
    "#print(train_labels.head())\n",
    "print(\"# of seqs with tails in train_seqs\",sum(df.loc[train_seqs.index].TAILLEN > 0))\n",
    "print(\"# of seqs with tails in test_seqs\",sum(df.loc[test_seqs.index].TAILLEN > 0))\n",
    "print(\"\\nThe following ratios should be as similar as possible (Adjust random seed accordingly):\")\n",
    "print(\"Train seqs/tails in train seqs =\",train_seqs.shape[0]/sum(df.loc[train_seqs.index].TAILLEN > 0))\n",
    "print(\"Test seqs/tails in test seqs =\",test_seqs.shape[0]/sum(df.loc[test_seqs.index].TAILLEN > 0))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9a2c7bd",
   "metadata": {},
   "source": [
    "<h3>Build model on train set</h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "id": "464d7ce4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[9.99520333e-01 4.79667431e-04 0.00000000e+00 0.00000000e+00]\n",
      " [0.00000000e+00 1.00000000e+00 0.00000000e+00 0.00000000e+00]\n",
      " [5.00000000e-01 5.00000000e-01 0.00000000e+00 0.00000000e+00]\n",
      " [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00]]\n"
     ]
    }
   ],
   "source": [
    "### Build first model and write to file as json object\n",
    "model_0 = train_hmm(distribution=DiscreteDistribution,\n",
    "                    n_components=len(states),\n",
    "                    state_names=states,\n",
    "                    X=[list(s) for s in list(train_seqs)], # Needs sequences as lists of characters\n",
    "                    labels=[list(s) for s in list(train_labels)], # Needs sequences as lists of characters\n",
    "                    algorithm='labeled',\n",
    "                    n_jobs=2)\n",
    "\n",
    "model_0_json = model_0.to_json()\n",
    "with open(\"model_0.json\",'w') as f:\n",
    "    f.write(model_0_json)\n",
    "\n",
    "#print(model_1)\n",
    "print(model_0.dense_transition_matrix())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b60cc3f",
   "metadata": {},
   "source": [
    "<h3>Model validation with grid search on transition probability matrix (Takes long time)</h3> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "id": "f24c921d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.9995203325694185 0.0004796674305814636 0.0 1.0\n",
      "FF_start = 0.9995203325694185         FF_end = 0.69\n",
      "TT_start = 0.999         TT_end = 0.79\n",
      "Step = -0.2\n"
     ]
    }
   ],
   "source": [
    "### Grid search on transition probability matrix is needed because there is no T to F transition in this data.\n",
    "### However, there could be T to F transition in new data.\n",
    "### Therefore, T to F transition probability has to be non-zero.\n",
    "### Grid search will allow to select a good enough model with non-zero T to F transition probability.\n",
    "\n",
    "### Also, the selected model at this stage has to be very precise in identifying tails.\n",
    "### Since tails are encoded as 1 and folded region as 0,\n",
    "### the calculated precision, TP/(TP+FP), gives precision of identifying tails.\n",
    "### Choose model with highest precision.\n",
    "\n",
    "### Read model from json file\n",
    "\n",
    "with open(\"model_0_PAKTO.json\", 'r') as f:\n",
    "    model_json = json.load(f)\n",
    "\n",
    "for a in model_json['edges']:\n",
    "    if (a[0],a[1]) == (0,0):\n",
    "        FF = a[2]\n",
    "    elif (a[0],a[1]) == (0,1):\n",
    "        FT = a[2]\n",
    "    elif (a[0],a[1]) == (1,0):\n",
    "        TF = a[2]\n",
    "    elif (a[0],a[1]) == (1,1):\n",
    "        TT = a[2]\n",
    "print(FF,FT,TF,TT)\n",
    "\n",
    "### Set state transition probabilites for grid search\n",
    "### Run this block multiple times adjusting/refining the size of the grid\n",
    "FF_start = model_json['edges'][2][2]\n",
    "FF_end = 0.69# 0.89\n",
    "print(\"FF_start =\",FF_start,\"        FF_end =\",FF_end)\n",
    "TT_start = 0.999\n",
    "TT_end = 0.79# 0.001\n",
    "print(\"TT_start =\",TT_start,\"        TT_end =\",TT_end)\n",
    "step = -0.2#-0.01# 0.1\n",
    "print(\"Step =\",step)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "id": "1bf4e173",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating 6 6\n",
      "Validating 6 6\n",
      "Validating 6 6\n",
      "Validating 6 6\n",
      "6\n"
     ]
    }
   ],
   "source": [
    "### Grid search for all combinations of FF&FT and TF&TT\n",
    "### Change file name for output of grid search each time\n",
    "#This block takes long time to run. Uncommented line to avoid executing this block accedentally.\n",
    "\n",
    "f = open(\"gridsearch_2_PAKTO.csv\",'w')\n",
    "f.write(\"FF,FT,TF,TT,Recall,Precision,Accuracy,MCC,AUC,F1score\\n\")\n",
    "for FF in np.arange(FF_start,FF_end,step):\n",
    "    FT = 1.0 - FF\n",
    "    for TT in np.arange(TT_start,TT_end,step):\n",
    "        TF = 1.0 - TT\n",
    "        #print('{:.5f}'.format(FF), '{:.5f}'.format(FT))\n",
    "        #print('{:.5f}'.format(TF), '{:.5f}'.format(TT))\n",
    "        #print()\n",
    "        \n",
    "        #model_json['edges'][2][2] = FF\n",
    "        #model_json['edges'][3][2] = FT\n",
    "        #model_json['edges'][4][2] = TF\n",
    "        #model_json['edges'][5][2] = TT\n",
    "\n",
    "        cnt = 0\n",
    "        for a in model_json['edges']:\n",
    "            if (a[0],a[1]) == (0,0):\n",
    "                model_json['edges'][cnt][2] = FF\n",
    "            elif (a[0],a[1]) == (0,1):\n",
    "                model_json['edges'][cnt][2] = FT\n",
    "            elif (a[0],a[1]) == (1,0):\n",
    "                model_json['edges'][cnt][2] = TF\n",
    "            elif (a[0],a[1]) == (1,1):\n",
    "                model_json['edges'][cnt][2] = TT\n",
    "            cnt += 1\n",
    "\n",
    "        model = HiddenMarkovModel.from_json(json.dumps(model_json, indent=4))\n",
    "        print(\"Validating\", cnt, len(model_json['edges']))\n",
    "        #val = validate(model=model, states=states, true_seqs=test_seqs, true_labels=test_labels)\n",
    "#         f.write(\",\".join([str(s) for s in [FF, FT, TF, TT,\n",
    "#                                            val[\"Recall\"], val[\"Precision\"],                                            \n",
    "#                                            val[\"Accuracy\"], val[\"MCC\"],\n",
    "#                                            val[\"AUC\"], val[\"F1score\"]]]) + \"\\n\")\n",
    "f.close()\n",
    "print(cnt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 275,
   "id": "b86a7d95",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Generate model with selected state transition probabilities\n",
    "FF = 0.9995203326\n",
    "FT = 0.0004796674\n",
    "TF = 0.201\n",
    "TT = 0.799\n",
    "\n",
    "with open(\"model_0.json\", 'r') as f:\n",
    "    model_json = json.load(f)\n",
    "\n",
    "cnt = 0\n",
    "for a in model_json['edges']:\n",
    "    if (a[0],a[1]) == (0,0):\n",
    "        model_json['edges'][cnt][2] = FF\n",
    "    elif (a[0],a[1]) == (0,1):\n",
    "        model_json['edges'][cnt][2] = FT\n",
    "    elif (a[0],a[1]) == (1,0):\n",
    "        model_json['edges'][cnt][2] = TF\n",
    "    elif (a[0],a[1]) == (1,1):\n",
    "        model_json['edges'][cnt][2] = TT\n",
    "    cnt += 1\n",
    "    \n",
    "model = HiddenMarkovModel.from_json(json.dumps(model_json, indent=4))\n",
    "with open(\"model_1.json\",'w') as f:\n",
    "    f.write(model.to_json())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24399cbf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80522ef4",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "8adadecd",
   "metadata": {},
   "source": [
    "<h3>Predict tail-like regions in new sequences using model_1</h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "id": "94f7daa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Read model\n",
    "model = HiddenMarkovModel.from_json('model_1.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "id": "ad40bbe9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 1353 new sequences.\n"
     ]
    }
   ],
   "source": [
    "### Get new sequences for which predictions will be generated\n",
    "\n",
    "filelist = [\n",
    "#'adk_cdhit90.fasta',\n",
    "'hbha_cdhit90.fasta',\n",
    "'hc1_cdhit90.fasta',\n",
    "'hc2_cdhit90.fasta',\n",
    "#'hu_cdhit90.fasta'\n",
    "'ku_cdhit90.fasta',\n",
    "#'l22_cdhit90.fasta',\n",
    "#'rbr_cdhit90.fasta',#? Manually identified tails are probably not correct PAK tails.\n",
    "#'sig_cdhit90.fasta',#? Manually identified tails are probably not correct PAK tails.\n",
    "#'tig_cdhit90.fasta',\n",
    "#'topo1_cdhit90.fasta',\n",
    "#'topo3_cdhit90.fasta'\n",
    "]\n",
    "new_seqs = get_seq_dictionary(datadir=datadir, filelist=full_seq_files, header_regex_str='>(.+?)\\s.*\\n', translation_dict=aa_code_cat_dict, verbose=True)\n",
    "\n",
    "#seq_tail = get_seq_dictionary(datadir=datadir, filelist=tail_seq_files, header_regex_str='>(.+).*\\n', translation_dict=aa_code_cat_dict, verbose=True)\n",
    "#seq_full = get_seq_dictionary(datadir=datadir, filelist=full_seq_files, header_regex_str='>(.+?)\\s.*\\n', translation_dict=aa_code_cat_dict, verbose=True)\n",
    "#print(\"Number of tail sequences:\",len(seq_tail),\"\\nNumber of all (full) sequences:\",len(seq_full))\n",
    "\n",
    "print(\"Found\",len(new_seqs),\"new sequences.\")\n",
    "#seqs = pd.Series(new_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "id": "b66318db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1353, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HEADER</th>\n",
       "      <th>START</th>\n",
       "      <th>END</th>\n",
       "      <th>LEN</th>\n",
       "      <th>SEQ</th>\n",
       "      <th>LABEL_SEQ</th>\n",
       "      <th>LABEL</th>\n",
       "      <th>DATASET</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tr|A0A064BZG4|A0A064BZG4_STREE</td>\n",
       "      <td>0</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>OAOKOOOOAKOAOATOOTKKOOAAAOOAOOAAOAOOOAAOOKOOOO...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "      <td>F</td>\n",
       "      <td>NEW_SEQS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sp|O33125|DBH_MYCLE</td>\n",
       "      <td>0</td>\n",
       "      <td>109</td>\n",
       "      <td>109</td>\n",
       "      <td>OOKAOOOOOOTOKOOOOOOOATAAOOOOOOTOOOAOOKOOOOTOTO...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "      <td>F</td>\n",
       "      <td>NEW_SEQS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sp|O33125|DBH_MYCLE</td>\n",
       "      <td>109</td>\n",
       "      <td>194</td>\n",
       "      <td>85</td>\n",
       "      <td>AAKKAAOKKAPOKKAOAKKAATKAPAKKAOKAPAKKOTTAOKOPAK...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT...</td>\n",
       "      <td>T</td>\n",
       "      <td>NEW_SEQS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sp|O33125|DBH_MYCLE</td>\n",
       "      <td>194</td>\n",
       "      <td>200</td>\n",
       "      <td>6</td>\n",
       "      <td>AKOOOK</td>\n",
       "      <td>FFFFFF</td>\n",
       "      <td>F</td>\n",
       "      <td>NEW_SEQS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sp|P0A1R8|DBHB_SALTY</td>\n",
       "      <td>0</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>OOKOOOOOKOAAOAOOOKAAAOOAOOAOOAOOTOOOKOOOOOAOOO...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "      <td>F</td>\n",
       "      <td>NEW_SEQS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           HEADER START  END  LEN  \\\n",
       "0  tr|A0A064BZG4|A0A064BZG4_STREE     0   91   91   \n",
       "1             sp|O33125|DBH_MYCLE     0  109  109   \n",
       "2             sp|O33125|DBH_MYCLE   109  194   85   \n",
       "3             sp|O33125|DBH_MYCLE   194  200    6   \n",
       "4            sp|P0A1R8|DBHB_SALTY     0   90   90   \n",
       "\n",
       "                                                 SEQ  \\\n",
       "0  OAOKOOOOAKOAOATOOTKKOOAAAOOAOOAAOAOOOAAOOKOOOO...   \n",
       "1  OOKAOOOOOOTOKOOOOOOOATAAOOOOOOTOOOAOOKOOOOTOTO...   \n",
       "2  AAKKAAOKKAPOKKAOAKKAATKAPAKKAOKAPAKKOTTAOKOPAK...   \n",
       "3                                             AKOOOK   \n",
       "4  OOKOOOOOKOAAOAOOOKAAAOOAOOAOOAOOTOOOKOOOOOAOOO...   \n",
       "\n",
       "                                           LABEL_SEQ LABEL   DATASET  \n",
       "0  FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...     F  NEW_SEQS  \n",
       "1  FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...     F  NEW_SEQS  \n",
       "2  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT...     T  NEW_SEQS  \n",
       "3                                             FFFFFF     F  NEW_SEQS  \n",
       "4  FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...     F  NEW_SEQS  "
      ]
     },
     "execution_count": 286,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Generate dataframe from new sequences\n",
    "\n",
    "#s = \"TTFFFFTTTFFFFFTFFTTTTT\"\n",
    "#print(s[0:2])\n",
    "df_new = pd.DataFrame(columns=['HEADER','START','END','LEN','SEQ','LABEL_SEQ','LABEL','DATASET'])\n",
    "\n",
    "for header in new_seqs:\n",
    "  seq = new_seqs[header]\n",
    "  pred = model.predict(list(seq))\n",
    "  pred_seq = \"\".join([states[i] for i in pred])\n",
    "  #seq = \"AABABDBADBBADBRRRRRR\"\n",
    "  #pred_seq = \"TTFFFFTTTFFFFFTFFTTTTT\"\n",
    "  #print(seq)\n",
    "  #print(pred_seq)\n",
    "\n",
    "  #print(len(pred_seq))\n",
    "  start = 0\n",
    "\n",
    "  for substr in [match[0] for match in re.findall(r'((\\w)\\2{0,})', pred_seq)]:\n",
    "    ### This regex also works: [match[0] for match in re.findall(r'((.)\\2{,})', pred_seq)]\n",
    "    #print(substr)\n",
    "    end = start+len(substr)\n",
    "    #print(start,end)\n",
    "    #print(header,start,end,seq[start:end],pred_seq[start:end],'NEW_SEQS')\n",
    "#     df_new = df_new.append(pd.DataFrame([(header, start, end, end-start,\n",
    "#                                                   seq[start:end], pred_seq[start:end],\n",
    "#                                                   pred_seq[start], 'NEW_SEQS')],\n",
    "#                                                 columns=['HEADER','START','END','LEN',\n",
    "#                                                          'SEQ','LABEL_SEQ','LABEL','DATASET']),\n",
    "#                                    ignore_index=True)\n",
    "\n",
    "    df_new = pd.concat([df_new,\n",
    "                        pd.DataFrame([(header, start, end, end-start,\n",
    "                                       seq[start:end], pred_seq[start:end],\n",
    "                                       pred_seq[start], 'NEW_SEQS')],\n",
    "                                       columns=['HEADER','START','END','LEN','SEQ','LABEL_SEQ','LABEL','DATASET']\n",
    "                                    )\n",
    "                       ],\n",
    "                       ignore_index=True)\n",
    "\n",
    "    start += len(substr)\n",
    "  #print(header,start,end,seq[start:end],pred_seq[start:],'NEW_SEQS')\n",
    "  #break\n",
    "\n",
    "print(df.shape)\n",
    "df_new.head()\n",
    "#new_seq_df[new_seq_df.LABEL == 'F']['LEN'].plot.hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "id": "59526d2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-288-4536cbf966ad>:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['DATASET'].iloc[train_seqs.index] = 'TRAIN'\n",
      "<ipython-input-288-4536cbf966ad>:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['DATASET'].iloc[test_seqs.index] = 'TEST'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Columns in df: Index(['FULL', 'TAIL', 'FULLLEN', 'TAILLEN', 'FOLDLEN', 'FULLLABELS', 'HEADER',\n",
      "       'DATASET'],\n",
      "      dtype='object')\n",
      "TRAIN    811\n",
      "TEST     542\n",
      "Name: DATASET, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HEADER</th>\n",
       "      <th>START</th>\n",
       "      <th>END</th>\n",
       "      <th>LEN</th>\n",
       "      <th>SEQ</th>\n",
       "      <th>LABEL_SEQ</th>\n",
       "      <th>LABEL</th>\n",
       "      <th>DATASET</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sp|O33125|DBH_MYCLE</td>\n",
       "      <td>0</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>OOKAOOOOOOTOKOOOOOOOATAAOOOOOOTOOOAOOKOOOOTOTO...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "      <td>F</td>\n",
       "      <td>TEST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sp|O33125|DBH_MYCLE</td>\n",
       "      <td>89</td>\n",
       "      <td>200</td>\n",
       "      <td>111</td>\n",
       "      <td>AOAOOOPOOOPAOKOOOATOAAKKAAOKKAPOKKAOAKKAATKAPA...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT...</td>\n",
       "      <td>T</td>\n",
       "      <td>TEST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sp|P02348|DBH5_RHILE</td>\n",
       "      <td>0</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>OOKOOOOOAOAOKAOOTKOOAAOAOOAOOOOOOAOOKOKOOOOOAO...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "      <td>F</td>\n",
       "      <td>TEST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sp|P05384|DBHB_PSEAE</td>\n",
       "      <td>0</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>OOKOOOOOAOAAOAOOPKAOAOOAOOAOOOOOTOAOKAOOOOOOOO...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "      <td>F</td>\n",
       "      <td>TEST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sp|P0A1R8|DBHB_SALTY</td>\n",
       "      <td>0</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>OOKOOOOOKOAAOAOOOKAAAOOAOOAOOAOOTOOOKOOOOOAOOO...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "      <td>F</td>\n",
       "      <td>TEST</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 HEADER START  END  LEN  \\\n",
       "0   sp|O33125|DBH_MYCLE     0   89   89   \n",
       "1   sp|O33125|DBH_MYCLE    89  200  111   \n",
       "2  sp|P02348|DBH5_RHILE     0   91   91   \n",
       "3  sp|P05384|DBHB_PSEAE     0   90   90   \n",
       "4  sp|P0A1R8|DBHB_SALTY     0   90   90   \n",
       "\n",
       "                                                 SEQ  \\\n",
       "0  OOKAOOOOOOTOKOOOOOOOATAAOOOOOOTOOOAOOKOOOOTOTO...   \n",
       "1  AOAOOOPOOOPAOKOOOATOAAKKAAOKKAPOKKAOAKKAATKAPA...   \n",
       "2  OOKOOOOOAOAOKAOOTKOOAAOAOOAOOOOOOAOOKOKOOOOOAO...   \n",
       "3  OOKOOOOOAOAAOAOOPKAOAOOAOOAOOOOOTOAOKAOOOOOOOO...   \n",
       "4  OOKOOOOOKOAAOAOOOKAAAOOAOOAOOAOOTOOOKOOOOOAOOO...   \n",
       "\n",
       "                                           LABEL_SEQ LABEL DATASET  \n",
       "0  FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...     F    TEST  \n",
       "1  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT...     T    TEST  \n",
       "2  FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...     F    TEST  \n",
       "3  FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...     F    TEST  \n",
       "4  FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...     F    TEST  "
      ]
     },
     "execution_count": 288,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### ADD a column to train/test data (i.e. Dataframe df) to hold dataset name\n",
    "df['DATASET'] = ''\n",
    "df['DATASET'].iloc[train_seqs.index] = 'TRAIN'\n",
    "df['DATASET'].iloc[test_seqs.index] = 'TEST'\n",
    "#print(df['DATASET'].value_counts())\n",
    "#print(train_seqs.shape,test_seqs.shape)\n",
    "#print(df[df['DATASET']=='TRAIN'].head())\n",
    "print(\"Columns in df:\",df.columns)\n",
    "print(df['DATASET'].value_counts())\n",
    "\n",
    "### Create temporary dataframe in required format\n",
    "df_train_test = pd.DataFrame(columns=['HEADER','START','END','LEN','SEQ','LABEL_SEQ','LABEL','DATASET'])\n",
    "\n",
    "for i in df.index:\n",
    "  #print(df.iloc[i])\n",
    "  row = df.iloc[i]\n",
    "  start = 0\n",
    "  for substr in [match[0] for match in re.findall(r'((\\w)\\2{0,})', row['FULLLABELS'])]:\n",
    "    #print(\"*\",substr)\n",
    "    end = start+len(substr)\n",
    "    #print(start,end)\n",
    "#     df_train_test = df_train_test.append(pd.DataFrame([(row['HEADER'], start, end, end-start,\n",
    "#                                                   row['FULL'][start:end], row['FULLLABELS'][start:end],\n",
    "#                                                   row['FULLLABELS'][start], row['DATASET'])],\n",
    "#                                                 columns=['HEADER','START','END','LEN',\n",
    "#                                                          'SEQ','LABEL_SEQ','LABEL','DATASET']),\n",
    "#                                    ignore_index=True)\n",
    "\n",
    "    df_train_test = pd.concat([df_train_test,\n",
    "                               pd.DataFrame([(row['HEADER'], start, end, end-start,\n",
    "                                              row['FULL'][start:end], row['FULLLABELS'][start:end],\n",
    "                                              row['FULLLABELS'][start], row['DATASET'])],\n",
    "                                              columns=['HEADER','START','END','LEN','SEQ','LABEL_SEQ','LABEL','DATASET']\n",
    "                                           )\n",
    "                              ],\n",
    "                              ignore_index=True)\n",
    "\n",
    "    start += len(substr)\n",
    "  \n",
    "df_train_test.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "id": "80af8eb8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['HEADER', 'START', 'END', 'LEN', 'SEQ', 'LABEL_SEQ', 'LABEL',\n",
      "       'DATASET'],\n",
      "      dtype='object')\n",
      "Index(['HEADER', 'START', 'END', 'LEN', 'SEQ', 'LABEL_SEQ', 'LABEL',\n",
      "       'DATASET'],\n",
      "      dtype='object')\n",
      "NEW_SEQS    1465\n",
      "TRAIN        847\n",
      "TEST         565\n",
      "Name: DATASET, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HEADER</th>\n",
       "      <th>START</th>\n",
       "      <th>END</th>\n",
       "      <th>LEN</th>\n",
       "      <th>SEQ</th>\n",
       "      <th>LABEL_SEQ</th>\n",
       "      <th>LABEL</th>\n",
       "      <th>DATASET</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sp|O33125|DBH_MYCLE</td>\n",
       "      <td>0</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>OOKAOOOOOOTOKOOOOOOOATAAOOOOOOTOOOAOOKOOOOTOTO...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "      <td>F</td>\n",
       "      <td>TEST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sp|O33125|DBH_MYCLE</td>\n",
       "      <td>89</td>\n",
       "      <td>200</td>\n",
       "      <td>111</td>\n",
       "      <td>AOAOOOPOOOPAOKOOOATOAAKKAAOKKAPOKKAOAKKAATKAPA...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT...</td>\n",
       "      <td>T</td>\n",
       "      <td>TEST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sp|P02348|DBH5_RHILE</td>\n",
       "      <td>0</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>OOKOOOOOAOAOKAOOTKOOAAOAOOAOOOOOOAOOKOKOOOOOAO...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "      <td>F</td>\n",
       "      <td>TEST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sp|P05384|DBHB_PSEAE</td>\n",
       "      <td>0</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>OOKOOOOOAOAAOAOOPKAOAOOAOOAOOOOOTOAOKAOOOOOOOO...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "      <td>F</td>\n",
       "      <td>TEST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sp|P0A1R8|DBHB_SALTY</td>\n",
       "      <td>0</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>OOKOOOOOKOAAOAOOOKAAAOOAOOAOOAOOTOOOKOOOOOAOOO...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "      <td>F</td>\n",
       "      <td>TEST</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 HEADER START  END  LEN  \\\n",
       "0   sp|O33125|DBH_MYCLE     0   89   89   \n",
       "1   sp|O33125|DBH_MYCLE    89  200  111   \n",
       "2  sp|P02348|DBH5_RHILE     0   91   91   \n",
       "3  sp|P05384|DBHB_PSEAE     0   90   90   \n",
       "4  sp|P0A1R8|DBHB_SALTY     0   90   90   \n",
       "\n",
       "                                                 SEQ  \\\n",
       "0  OOKAOOOOOOTOKOOOOOOOATAAOOOOOOTOOOAOOKOOOOTOTO...   \n",
       "1  AOAOOOPOOOPAOKOOOATOAAKKAAOKKAPOKKAOAKKAATKAPA...   \n",
       "2  OOKOOOOOAOAOKAOOTKOOAAOAOOAOOOOOOAOOKOKOOOOOAO...   \n",
       "3  OOKOOOOOAOAAOAOOPKAOAOOAOOAOOOOOTOAOKAOOOOOOOO...   \n",
       "4  OOKOOOOOKOAAOAOOOKAAAOOAOOAOOAOOTOOOKOOOOOAOOO...   \n",
       "\n",
       "                                           LABEL_SEQ LABEL DATASET  \n",
       "0  FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...     F    TEST  \n",
       "1  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT...     T    TEST  \n",
       "2  FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...     F    TEST  \n",
       "3  FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...     F    TEST  \n",
       "4  FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...     F    TEST  "
      ]
     },
     "execution_count": 289,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Concatanate dataframes for new sequences and train/test dataset\n",
    "print(df_train_test.columns)\n",
    "print(df_new.columns)\n",
    "\n",
    "df_2 = pd.concat([df_train_test, df_new], axis=0, ignore_index=True)\n",
    "\n",
    "print(df_2['DATASET'].value_counts())\n",
    "df_2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0756d06",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hmm-annotate-sequence",
   "language": "python",
   "name": "hmm-annotate-sequence"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
